Add External Source (and External Event) Attribute Validation

pranav-super opened the pull request:
```___REQUIRES_AERIE_PR___="1606"```

```___REQUIRES_GATEWAY_PR___="116"```

# Description
This PR adds external source and event attribute validation to AERIE.

It is explained more in depth [here](https://github.com/NASA-AMMOS/aerie/pull/1606). 

The main updates to the UI for this PR surround the displaying of source and event attributes in the UI, namely in the External Source Manager, as well as an external source/event type creation page (previously Modals). Derivation group creation has presently been moved out of a modal and is delegated to the `Derivation Group` field where sources are uploaded.

Below is a view of the new 'External Types Manager' with an External Source Type's attribute schema being shown:
![image](https://github.com/user-attachments/assets/49f929bb-94e1-4c90-925f-5fb721898e56)


Below is a view of an External Event's attributes being shown:
![image](https://github.com/user-attachments/assets/003c43c8-421e-4a67-a1d1-8658da6adeb1)

# Walkthrough

## Upload
To use this feature, navigate to the `External Sources` page of Aerie 

Select the `Types` button in the top-right to switch to the `/types` sub-page.

Click the 'Choose File' button and select a complete External Source & Event type schema (see bottom of this section for an example file). The pane should display the External Source & Event types that will be created once the 'Upload' button has been clicked - click the 'Upload' button to submit & create the types.

After the types have been created, a source can be uploaded by clicking the `Sources` button in the top-right, then selecting the `Choose File` button in the left-pane, selecting the related External Source, and clicking `Upload`.

## Viewing Schemas

The attribute schema for an uploaded External Event or Source Type can be viewed by navigating to the `External Sources` page and selecting the `Types` button in the top-right to switch to the `/types` sub-page.

Once on the `Types` page, select the tab for the desired type (`Event` or `Source`) and then hover the row containing the desired type. Click the `View` button that appears. On the left-pane, the type should be displayed with two collapsible categories for `Attribute Schema - Definition` and `Attribute Schema - Properties`. `Definition` contains any top-level fields included other than `Properties`, which is shown under the other collapse.

## Viewing Attributes

The attributes of a given External Source or Event can be seen by navigating to the `External Sources` page, hovering over the row for the desired External Source, and clicking the `View` button. In the `Selected External Source` pane on the left, there is a section for `Attributes` which when clicked will show the specific attributes for the given source.

## Example Files

**Example External Source**: 
[WEATHER_2024001_2024008_IDEAL.V00.json](https://github.com/user-attachments/files/21004821/WEATHER_2024001_2024008_IDEAL.V00.json)


**Example External Source Schema**: 
[weather-schema.json](https://github.com/user-attachments/files/21004818/weather-schema.json)

JosephVolosin committed:
Prototype of creating external source/event types thru Gateway

JosephVolosin committed:
Prototype WIP

JosephVolosin committed:
WIP Prototype of visualizing attributes

Commit:
update to play better with gateway and source validation

JosephVolosin committed:
Prototype cleanup, re-adding permissions, removing code that is now in Gateway

JosephVolosin committed:
Lint/format cleanup

Commit:
fix Uploading... bug

Commit:
additional error handling

Commit:
remove allowed event types from source type creation

Commit:
finish error handling

Commit:
complete error handling

JosephVolosin committed:
Rework JsonSchema to translate to ValueSchema for UI

JosephVolosin committed:
Cleanup

JosephVolosin committed:
Use <Parameters> in Management modal

JosephVolosin committed:
Remove extra handling for errors

JosephVolosin committed:
Close creation modal after successful creation

JosephVolosin committed:
handle sources with some defined and some undefined types more clearly, improve error handling

JosephVolosin committed:
Add attributes to expected test output, fix formatting

JosephVolosin committed:
Error handling fixes & e2e test updates

JosephVolosin committed:
Improve handling for non-existent types

JosephVolosin committed:
Lint fix

JosephVolosin committed:
Refactor createExternalSource to send parsed JSON as form data

Commit:
update effects for new source type uploading system

JosephVolosin committed:
WIP - Update modal for creating types from a single schema

JosephVolosin committed:
Rework Create & Manage modals into one, fix bug with ExternalEventForm

ExternalEventForm's calculated attributes for the given event are now reactive to fix a bug where attributes didn't update when the selected event was switched

JosephVolosin committed:
Move external type management modal to sub-page of External Sources

JosephVolosin committed:
Update e2e tests for new 'type' page

JosephVolosin committed:
Updating derivation examples for e2e tests

JosephVolosin committed:
--amend

JosephVolosin committed:
Add support for arrays as series

JosephVolosin committed:
Fix issue with auto-selecting source after upload

JosephVolosin committed:
Formatting

JosephVolosin committed:
Always show all 3 tables in type manager

JosephVolosin commented:
Having some issues with the E2E tests, but all External Source tests are passing locally. Note that because there is a separate Gateway branch for this pull-request there's issues with running the Gateway branch locally causing model upload, I worked around this by forcing it to use a pre-uploaded model to confirm that the actual tests still pass.

`e2e-tests/external-sources.test.ts`
![image](https://github.com/user-attachments/assets/3b913cbe-8ca2-48ec-b933-ed93835b8901)
 
`e2e-tests/plan-external-sources.test.ts`
![image](https://github.com/user-attachments/assets/e5e9d1f0-553e-439f-a1a9-e305ccff568d)

JosephVolosin committed:
Linting & test fixes

JosephVolosin committed:
Formatting on ExternalTypeManager, use Panels like Dictionaries page instead of CSSGridGutters

JosephVolosin committed:
Update formatting for Arrays in attribute schemas, add associated sources to event types

JosephVolosin committed:
Formatting

JosephVolosin committed:
Properly translate nested JsonSchema-typed items to ValueSchema

JosephVolosin committed:
Add missing deselectSource

JosephVolosin committed:
E2E test fixes

Test lenght is becoming a concern, so tests may be flaky in this commit

JosephVolosin committed:
Remove out-of-date comment

JosephVolosin committed:
Fix typo; names => name

Commit:
update to allow automatic creation of types and allow either/or event/source types in schema

Commit:
show required attributes

JosephVolosin committed:
Fix bug with not displaying objects/structs nested within arrays/series

Commit:
update tests

duranb commented on src/utilities/effects.ts:
```diff
@@ -999,69 +977,109 @@ const effects = {
-    sourceType: ExternalSourceTypeInsertInput,
-    user: User | null,
-  ): Promise<ExternalSourceType | undefined> {
+  async createExternalSourceEventTypes(eventTypes: object | undefined, sourceTypes: object | undefined, user: User | null) {
```
Can we define the return type for this and `createExternalSource` above?

JosephVolosin replied:
Added return types to both `createExternalSource` and `createExternalSourceEventTypes`

duranb commented on src/utilities/effects.ts:
```diff
@@ -999,69 +977,109 @@ const effects = {
+        throw new Error(respErrors);
       }
+      creatingExternalSource.set(false);
+      return reqResponse;
```
Can this return the object in the response rather than the response itself? It'll be easier to use if this function does the checking of whether `createExternalSource` exists in the response rather than within the components like `ExternalSourceManager.svelte`

JosephVolosin replied:
Yep - moved the deconstruction of the response into here so the function now returns an `ExternalSourceSlim` upon success

duranb commented on src/utilities/effects.ts:
```diff
@@ -863,31 +858,39 @@ const effects = {
   },
 
-  async createExternalEventType(eventType: ExternalEventTypeInsertInput, user: User | null) {
+  async createExternalEventType(eventTypeName: string, eventTypeAttributesSchema: File, user: User | null) {
```
Can we define a return type for this?

JosephVolosin replied:
Think we forgot to cleanup this route earlier - we no longer use `createExternalSourceType`/`createExternalEventType` since we condensed to a single route on Gateway for `createExternalSourceEventType` (so the user can upload a single file with source AND event types rather than multiple files). Removed the function!

duranb commented on src/utilities/effects.ts:
```diff
@@ -999,69 +977,109 @@ const effects = {
       } else {
-        throw Error(`Unable to create external source type`);
+        showFailureToast('External Source & Event Type Create Failed');
+        console.log(response.errors)
```
Leftover log

duranb commented on src/utilities/effects.ts:
```diff
@@ -999,69 +977,109 @@ const effects = {
+        source_type_name: externalSourceTypeName,
+        valid_at: validAtFormatted,
+      };
+      const body = new FormData();
```
This doesn't have to be a `FormData` object. We only use `FormData` for cases when we have files being uploaded. Otherwise, this can just stay a simple JS object that's stringified before going into `reqGateway`

JosephVolosin replied:
With sending as `application/json`, very large files seem to overflow the request:
<img width="1508" alt="Screenshot 2025-04-11 at 10 53 43 AM" src="https://github.com/user-attachments/assets/76185a79-c5f8-4daf-ab2b-d3c46a6a76e7" />
The input file here is ~30MB and something we'd expect to be able to upload - maybe the answer is just having the input split up prior to uploading to Aerie though?

duranb replied:
What file is being uploaded here?

JosephVolosin replied:
It's a 30MB JSON from one of our GSFC missions - I can make a 'generic' version of it if you want to test also

duranb replied:
Sorry, what I meant is that this function doesn't take a file argument to indicate the need for using a FormData to upload a file. What argument is 30mb big?

JosephVolosin replied:
Ah - so we _are_ "uploading" a file which represents the external source, but only the parsed content is sent to Gateway (i.e., no file is "uploaded" and **saved** in the database). The "30MB" argument is the parsed and formatted content of the 30MB file (i.e., the external source being "uploaded")

duranb replied:
Is it possible to upload the file and then have the gateway parse it? That's how we currently do it for plan imports.

JosephVolosin replied:
It should be possible, I think we avoided it in the past because it auto-fills data once the file has been selected but not yet uploaded so the user can see the key, derivation group, start/stop times, etc. for the file. I don't think anything is necessary though as far as why it's parsed on the front-end vs. Gateway

duranb replied:
Ah okay. If you don't mind, can you change it to upload the file to the gateway and then have the gateway parse it and make queries to hasura to insert? For plan imports, we have the gateway split up the queries into several requests to hasura for this very purpose of avoiding huge payload sizes. The gateway is also in charge of cleaning up in case any of the requests failed so that there isn't any lingering incomplete artifacts in the db.

JosephVolosin replied:
Swapped the parsing logic to Gateway now! The external source upload now sends the file as `multipart/form-data` (with an optional Derivation Group field), and the type upload is now sent as generic `application/json` instead of `multipart/form-data` (size constraints shouldn't apply to those). See the following commits on `aerie-gateway`:

* https://github.com/NASA-AMMOS/aerie-gateway/pull/116/commits/768bb8eedca6e96f1ab80e5d896b590685c5d69c
* https://github.com/NASA-AMMOS/aerie-gateway/pull/116/commits/6602dc99e4bb74a91d1705a2d2a54fcdcf92eefd
* https://github.com/NASA-AMMOS/aerie-gateway/pull/116/commits/d2fbab7734b619f1523b21058966c7ca22fbb799

duranb commented on src/utilities/effects.ts:
```diff
@@ -999,69 +977,109 @@ const effects = {
   },
 
-  async createGroupsOrTypes(user: User | null): Promise<void> {
+  async createExternalSourceType(sourceTypeName: string, sourceTypeAttributesSchema: File, user: User | null) {
```
Can we define a return type for here as well?

JosephVolosin replied:
Ended up removing this function, see comment on the other related function/route: https://github.com/NASA-AMMOS/aerie-ui/pull/1560#discussion_r2038138792

duranb commented on src/utilities/effects.ts:
```diff
@@ -3718,11 +3736,13 @@ const effects = {
       const { external_source } = data;
       if (external_source != null) {
-        const event_types: string[] = [];
+        const event_types: ExternalEventType[] = [];
```
Can this be camelCased?

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -321,34 +374,75 @@
+          ? await effects.createExternalSourceEventTypes(newEventTypes, newSourceType, user)
+          : true;
+
+      console.log('CREATED TYPES: ', createdTypes);
```
leftover log

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -321,34 +374,75 @@
+          // Auto-select the new source
+          selectedSource = {
+            ...createExternalSourceResponse,
+            created_at: new Date().toISOString().replace('Z', '+00:00'), // technically not the exact time it shows up in the database
```
`replace('Z', '+00:00')` seems to be being used in a few more places now. Would you mind splitting this out into a time utility function?

JosephVolosin replied:
Added `switchISOTimezoneRepresentation` to go both ways between `Z` and the `+00:00` representation!

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -408,21 +502,51 @@
+        newSourceExternalEventTypes[event.event_type] = Object.keys(event.attributes);
+      } else if (newSourceExternalEventTypes[event.event_type] !== Object.keys(event.attributes)) {
+        // if there is a mismatch, cause error
+        console.log(newSourceExternalEventTypes[event.event_type], Object.keys(event.attributes));
```
leftover log

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -408,21 +502,51 @@
+      }
+    }
+    if (eventTypes.length > 0) {
+      console.log(externalSource.events);
```
leftover log

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -302,6 +342,18 @@
+    }
+  } else {
+    isUploadDisabled = true;
+    uploadDisabledMessage = null;
```
Should this be set to `null` when `isUploadDisabled` becomes `true`?

duranb replied:
It doesn't seem to be being set to anything at all actually.

JosephVolosin replied:
Added some actual use for `uploadDisabledMessage` when the UI does quick checks before sending to Gateway

duranb commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -408,21 +502,51 @@
+    }
+
+    let eventTypes: string[] = [];
+    for (const entry of Object.entries(newSourceExternalEventTypes)) {
```
Do you mind if this is rewritten as something like `for (const [eventType, eventAttributes] Object.entries(...)) {}` so that the following code can be a little clearer?

duranb commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,993 @@
+  let externalSourceTypeColumnDefs: DataGridColumnDef<ExternalSourceType>[] = externalSourceTypeBaseColumnDefs;
+  let externalEventTypeColumnDefs: DataGridColumnDef<ExternalEventType>[] = externalEventTypeBaseColumnDefs;
+
+  let derivationGroupDataGrid: DataGrid<DerivationGroup>;
```
Are these being used within this component?

JosephVolosin replied:
Nope - removed the unused `DataGrid` variables

duranb commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,993 @@
+        throw new Error('External Source & Event Type Schema has Invalid Format');
+      }
+    } catch (error) {
+      createExternalSourceEventTypeError.set('External Source & Event Type Schema has Invalid Format');
```
should this just use the `error` being passed into the `catch` sine it's the same as the error being thrown?

JosephVolosin committed:
Formatting fix

JosephVolosin committed:
Add null condition for source/event types to be created

JosephVolosin committed:
Remove console.logs

JosephVolosin committed:
Fix comparison between event attribute keys during upload

JosephVolosin committed:
Format fix

AaronPlave commented:
When I select the JSON schema file given in the example I see a message that it parsed successfully but then when I go to upload it I see this error:
<img width="355" alt="image" src="https://github.com/user-attachments/assets/481280c3-bb55-4eb4-89bd-2637bb086fb7" />

AaronPlave commented:
@JosephVolosin @pranav-super  Seeing this error when uploading the sample source file after uploading the types:
>Internal Server Error {"message":"The source and/or event types in your source do not exist in the database."}

Any idea if I'm doing something wrong?

pranav-super commented:
> @JosephVolosin @pranav-super Seeing this error when uploading the sample source file after uploading the types:
> 
> > Internal Server Error {"message":"The source and/or event types in your source do not exist in the database."}
> 
> Any idea if I'm doing something wrong?

Hey! I'm not entirely sure. The only thing I can think of is maybe not all event types were uploaded? 
I'm assuming you're using the relevant branch of [aerie-gateway](https://github.com/NASA-AMMOS/aerie-gateway/pull/116) and [aerie](https://github.com/NASA-AMMOS/aerie/pull/1606)?

JosephVolosin committed:
Fix linting

JosephVolosin committed:
Fix svelte-check fail

Set.intersection() isn't supported in the current version of Svelte/TS

JosephVolosin committed:
Convert ExternalTypeManager DataGrids to BulkActionDataGrids

AaronPlave commented:
> > @JosephVolosin @pranav-super Seeing this error when uploading the sample source file after uploading the types:
> > > Internal Server Error {"message":"The source and/or event types in your source do not exist in the database."}
> > 
> > 
> > Any idea if I'm doing something wrong?
> 
> Hey! I'm not entirely sure. The only thing I can think of is maybe not all event types were uploaded? I'm assuming you're using the relevant branch of [aerie-gateway](https://github.com/NASA-AMMOS/aerie-gateway/pull/116) and [aerie](https://github.com/NASA-AMMOS/aerie/pull/1606)?

Yep, using relevant branches for aerie and gateway. What is weird is that the sample schema upload works and I see this:
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/dcf977ee-8b36-4adf-81fd-eebbdbcc324c" />

but then the subsequent example event source file upload fails with that message.

Update: Managed to get it working with one of the UI e2e sample external event files, perhaps something is awry with the sample one linked in the PR?

AaronPlave commented on src/components/external-events/ExternalEventForm.svelte:
```diff
@@ -65,6 +102,19 @@
           />
         </Input>
       </Collapse>
+      <Collapse defaultExpanded={false} title="Attributes" tooltipContent="View External Event Attributes">
```
Can you show something like ```<div class="st-typography-label">No Attributes Found</div>``` if none are present?

JosephVolosin replied:
Added!

AaronPlave replied:
Looks good, could you also add it for selected external source attributes?
<img width="297" alt="image" src="https://github.com/user-attachments/assets/bdc27b3c-57a0-4bf1-98af-1077978d363f" />

JosephVolosin replied:
Added!

AaronPlave commented on src/components/external-events/ExternalEventForm.svelte:
```diff
@@ -65,6 +102,19 @@
       </Collapse>
+      <Collapse defaultExpanded={false} title="Attributes" tooltipContent="View External Event Attributes">
+        <div class="st-typography-body">
+          <Parameters
```
Is the source override indicator that comes with `Parameters` used here?
<img width="716" alt="image" src="https://github.com/user-attachments/assets/8fe44459-4b50-4c6b-a8ed-67db99fb4c35" /> If not could you instead hide it via prop? Feel like it would cause confusion otherwise since no overrides are happening here right?

JosephVolosin replied:
It is the same piece of `Parameters` adding that - I updated `getFormParameters` to let you 'ignore' setting `ValueSource` which removes this. Let me know what you think!

AaronPlave replied:
Seems reasonable - other solution could I guess be passing down some sort of flag to not render the value source badge but that'd be a lot of prop drilling into all of the Parameter subcomponents so this is probably the cleanest way.

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1064 @@
+            {/if}
+          {/each}
+        </Collapse>
+        <Collapse
```
Can you show something like <div class="st-typography-label">No Properties Found</div> if none are present?

JosephVolosin replied:
Added to both the definition and properties sections!

AaronPlave replied:
Looks good! Could you also add it to the event type attribute schema - properties/definition?
<img width="1229" alt="image" src="https://github.com/user-attachments/assets/301706da-9aa4-4d82-8507-b0204a4cc1f6" />

JosephVolosin replied:
Added!

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1064 @@
+              </svelte:fragment>
+              <div>
+                <div class="st-typography-bold">Name:</div>
+                {associatedDerivationGroup.name}
```
Missing type styles on the name bit

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1064 @@
+      </svelte:fragment>
+      <svelte:fragment slot="body">
+        <div class="derivation-group-table">
+          <BulkActionDataGrid
```
I get a bit confused when I'm able to select one item from each table (derivation, external source types, and external event types) even though only one of those, i.e. the last clicked, shows up as selected on the left.
<img width="1474" alt="image" src="https://github.com/user-attachments/assets/02abef5c-159d-4ff5-ba73-2bd2af74f7c4" />

Is there any reason to leave the previously selected items highlighted? Do they filter the tables below in any way? If not I'd recommend trying to clear the selection on old rows in other tables to avoid confusion if possible.

AaronPlave replied:
Would also recommend clearing selection when the user closes the item details panel on the left.

JosephVolosin replied:
Fixed all these - you're now only able to select a single item (from _all tables_) at once. Close button also works properly now!

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
Noticing that after clicking on a row in any of the tables I can no longer resize the left panel

JosephVolosin replied:
Looks like this was related to replicating the `Panel` and `svelte:fragment` sections in multiple if/elif/else statements. Moved the conditionals to be under a single use of `Panel`/`svelte:fragment (body + header)` and it looks to be working now!

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1064 @@
+          tooltipContent={`${selectedExternalSourceType.name} Attribute Schema Definition`}
+          defaultExpanded={false}
+        >
+          {#each Object.entries(selectedExternalSourceType.attribute_schema) as attribute}
```
Could these either be left aligned within their column or stacked like other bits of text in this area? We don't really have any other right aligned text in aerie and I find it a bit difficult to read
<img width="345" alt="image" src="https://github.com/user-attachments/assets/faf5a7a2-21aa-4379-8079-50066c624dfb" />

JosephVolosin replied:
Removed the right-formatting so the values in the 'Definition' section are underneath their 'keys'

AaronPlave replied:
Hm, still seeing this
<img width="330" alt="image" src="https://github.com/user-attachments/assets/c56033a0-680d-4460-acc0-101aa7ed46dc" />

JosephVolosin replied:
Think this was breaking b/c of the unused stylings from `ExternalSourceManager` that were globally scoped - fixed in another comment, let me know if you're still seeing issues with this!

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1064 @@
+            hasDeletePermission: hasDeleteDerivationGroupPermissionOnRow(user, params.data),
+            rowData: params.data,
+            viewCallback: params.viewDerivationGroup,
+            viewTooltip: {
```
Does the view button do anything different from clicking on the row? If not, it feels like it is somewhat redundant and we could just have the row click be the only action?

JosephVolosin replied:
Good point - remove the view button as clicking the row is the same action

JosephVolosin committed:
Show 'No Attributes Found' if no Event attributes are present

JosephVolosin committed:
Add option to ignore valueSource for Parameters

JosephVolosin committed:
Remove 'view' button from ExternalTypeManager (and view fixes)

JosephVolosin committed:
Condense if/elif/else structure for header/body sections

JosephVolosin committed:
Show 'No Attributes Found' in definition if no Event attributes are present

JosephVolosin committed:
Add missing typography class

JosephVolosin committed:
Remove right-formatting from attribute definition

JosephVolosin committed:
Fix formatting

JosephVolosin commented:
> > > @JosephVolosin @pranav-super Seeing this error when uploading the sample source file after uploading the types:
> > > > Internal Server Error {"message":"The source and/or event types in your source do not exist in the database."}
> > > 
> > > 
> > > Any idea if I'm doing something wrong?
> > 
> > 
> > Hey! I'm not entirely sure. The only thing I can think of is maybe not all event types were uploaded? I'm assuming you're using the relevant branch of [aerie-gateway](https://github.com/NASA-AMMOS/aerie-gateway/pull/116) and [aerie](https://github.com/NASA-AMMOS/aerie/pull/1606)?
> 
> Yep, using relevant branches for aerie and gateway. What is weird is that the sample schema upload works and I see this: <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/4419607/407525877-dcf977ee-8b36-4adf-81fd-eebbdbcc324c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxODA4NDEsIm5iZiI6MTczODE4MDU0MSwicGF0aCI6Ii80NDE5NjA3LzQwNzUyNTg3Ny1kY2Y5NzdlZS04YjM2LTRhZGYtODFmZC1lZWJiZGJjYzMyNGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjlUMTk1NTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njk2MWI4YTU0ODI4OGZmOWU0YjM0YmE4MDAxNzRkOGJlNDJlMzBhZDRhOGU3NWY5NjE5ZTJmNzA5ZjY1YjdhNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.IX-Vx5JoyUggjs451e-TaOvn_WO71upkTLfEc-4-ogU">
> 
> but then the subsequent example event source file upload fails with that message.
> 
> Update: Managed to get it working with one of the UI e2e sample external event files, perhaps something is awry with the sample one linked in the PR?

Fixed the example products - they were a bit out-of-date with our latest schema work!

JosephVolosin committed:
Fix e2e tests

AaronPlave commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -321,34 +374,73 @@
-          ...createExternalSourceResponse,
-          created_at: new Date().toISOString().replace('Z', '+00:00'), // technically not the exact time it shows up in the database
+      // Create non-existing types first:
+      // TODO: cleanup
```
Any further action needed here?

JosephVolosin replied:
Nope! Removed the comment

AaronPlave commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1006 @@
+                Source & Event Type Attribute Schema Parsed
+              </div>
+            {:else}
+              <WarningIcon />
```
Can you wrap this in a div and align the icon and text together?
<img width="406" alt="image" src="https://github.com/user-attachments/assets/2f912e4d-e9e8-4b90-accd-fdcf1e9a2868" />

JosephVolosin replied:
Should be fixed now! The warning icon is also now properly colored red similar to other instances of it showing up throughout the UI

AaronPlave commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -588,6 +747,12 @@
           <AlertError class="m-2" error={$createExternalSourceError} />
           <AlertError class="m-2" error={$parsingError} />
+          <AlertError class="m-2" error={uploadDisabledMessage} />
+          <ExternalSourceTypeNewCard
```
How can I get this to show up?

JosephVolosin replied:
You should be able to get this to show up now - fixed a few related bugs w. it's conditions to show up as well as some related selection/de-selection stuff. See [this comment](https://github.com/NASA-AMMOS/aerie-ui/pull/1560#discussion_r1949716637) for more info on the use of this card!

AaronPlave commented on src/components/external-source/ExternalSourceManager.svelte:
```diff
@@ -790,6 +931,27 @@
 </CssGrid>
 
 <style>
+  :global(.attribute-name) {
```
Can these globals be scoped a bit more tightly under some parent class name?

JosephVolosin replied:
Not sure why these were in here - all the global stylings in this component were unused so I went ahead and removed them

AaronPlave commented on src/components/external-source/ExternalSourceTypeNewCard.svelte:
Just curious about how this card comes into play and what the purpose is - haven't managed to trigger it on my own yet.

JosephVolosin replied:
This card is supposed to show you the new External Source Types & Event Types that'll be created if you're uploading a source that doesn't have either/both defined. However, with attribute schemas we decided the only time the system should auto-generate types is if the source being uploaded doesn't contain any attributes for the undefined types. To trigger it, just set the `attributes` field to `{}` for both the source and event(s) and as long as you don't have schemas previously uploaded for the types it should trigger. Attaching an example below as well!

Note there was a few bugs with this feature that I fixed, refer to the other comment [here](https://github.com/NASA-AMMOS/aerie-ui/pull/1560#discussion_r1943367413)

[EmptyAttributesExample.json](https://github.com/user-attachments/files/18739392/EmptyAttributesExample.json)

AaronPlave commented on src/stores/external-event.ts:
```diff
@@ -11,8 +11,8 @@ import { gqlSubscribable } from './subscribable';
 /* Writeable. */
-export const creatingExternalEventType: Writable<boolean> = writable(false);
-export const createExternalEventTypeError: Writable<string | null> = writable(null);
+export const creatingExternalEventType: Writable<boolean> = writable(false); // TODO: remove
```
Still a TODO?

AaronPlave replied:
Seems to no longer be used

JosephVolosin replied:
Yep, no longer used!

JosephVolosin committed:
Remove old TODO

JosephVolosin committed:
Fix formatting on External Type Schema parse status

JosephVolosin committed:
Fix new type card & related bugs

JosephVolosin committed:
Remove unused global stylings; add 'No Attributes' found on event types

JosephVolosin committed:
Remove TODOs

JosephVolosin committed:
Move error store resetting to new function

AaronPlave reviewed (approved):
LGTM, apologies for the very drawn out review!

Commit:
update post-rebase

Commit:
cleanup

JosephVolosin committed:
Fix linting, tests

JosephVolosin committed:
Remove bad check for valid event types upon upload

The original check would enforce that each event-to-be-uploaded's attributes matched those of the other events-to-be-uploaded which is NOT a valid check as the user is allowed to upload a source where each event can contain different sets of attributes, so long as they are properly defined in the schema.

duranb commented on src/utilities/parameters.ts:
```diff
@@ -150,3 +155,139 @@ export function getCleansedStructArguments(structArguments: ArgumentsMap, schema
+      };
+      const propTranslated = translateJsonSchemaTypeToValueSchema(propType as JSONType, propProperties, propItems);
+      if ('items' in propTranslated) {
+        if (propTranslated.items.type === 'struct') {
```
It looks like the logic for both of these conditions are the same. Is there any way we can reduce the duplicate code?

JosephVolosin replied:
Reduced the actual logic down so that its only in there once - the only separated parts are referencing `propTranslated.items` vs `propTranslated.items.items`

JosephVolosin committed:
Remove unused routes for uploading types

JosephVolosin committed:
Update response for createExternalSource

JosephVolosin committed:
Remove old log

JosephVolosin committed:
Camel-case

JosephVolosin committed:
Add uploadDisabledMessage

JosephVolosin committed:
Updated error message for failed parse

JosephVolosin committed:
Reduce redundant code

duranb commented on src/components/external-source/ExternalTypeManager.svelte:
```diff
@@ -0,0 +1,1024 @@
+        throw new Error('External Source & Event Type Schema has Invalid Format');
+      }
+    } catch (error: any) {
+      createExternalSourceEventTypeError.set(error);
```
`error` in this case is an `Error` object, not a `string`. To access the error message to set the store to, you need to do 
```
createExternalSourceEventTypeError.set((error as Error).message);
```

JosephVolosin committed:
Fix error store set

JosephVolosin committed:
Removed unused datagrid vars

JosephVolosin committed:
Create function for swapping between timezone formats for ISO 8601 times

JosephVolosin committed:
Lint fix

JosephVolosin committed:
Move external source parsing to Gateway

JosephVolosin committed:
Send type schemas as JSON

JosephVolosin committed:
Remove unused e2e tests

No longer needed with parsing on the Gateway

JosephVolosin committed:
w-100 => w-full

JosephVolosin committed:
Fix lint

JosephVolosin committed:
Fix broken e2e test; add 'optional' to derivation group input

JosephVolosin committed:
Linting fixes

JosephVolosin committed:
Add missing 'derivation_group_name' field to example sources

dandelany commented:
Small issue discovered while testing with @JosephVolosin (not a dealbreaker but should fix if possible):

On the External Event **Types** page, under the "External Events Types" table, the "Associated External Sources" column value is funky and sometimes wrong. Eg. if you upload a type/source, then the column may show `0`, but if you click to the sources page and back, it will show the correct number

also - @JosephVolosin will add an e2e test case for uploading source/event with *missing* attributes key (as opposed to empty attributes)

JosephVolosin committed:
New E2E test for missing (optional) attributes field

JosephVolosin committed:
Fix broken association numbers on type manager

JosephVolosin committed:
Change 'optional' text for derivation group
